### PR TITLE
[AI Chat] Retrieve user's chat message and return Open AI response

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR retrieves the user input after they click on submit, adds the message to the existing conversation between the user and AI assistant which is then sent to the Open AI backend controller. When the response is received, the user input and assistant response are both rendered in the Chat Workspace.

The system prompt will be pulled from level properties in follow-up work and is simply hard-coded for now. Note that there is no filtering for personal information nor code of conduct violations yet.

Here is a screencast video - the system prompt is that the bot assistant acts as George Washington and is conversing with a middle school classroom. 

https://github.com/code-dot-org/code-dot-org/assets/107423305/a34a6a0a-6312-4acd-b937-3cbe9df2eca5

## Links
[jira](https://codedotorg.atlassian.net/browse/SL-1068)

## Follow-up work
- A user can only access Open AI backend controller with the `AI_CHAT_ACCESS` permission type. If they do not have this permission, return an appropriate error response.
- Add auto-scrolling once messages go beyond height of workspace - [jira](https://codedotorg.atlassian.net/browse/SL-1049)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/aichat/constants.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export const Role = {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/constants.ts]: It might be worth turning these into [enums](https://www.typescriptlang.org/docs/handbook/enums.html). You can still retain the string mapping by adding values, like
```
enum Role {
  ASSISTANT = 'assistant';
  USER = 'user';
...
```
but then you can refer to this enum as a type in other interfaces (`role: Role`)

File: apps/src/aichat/openai.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import {ChatCompletionMessage} from './types';
+import HttpClient from '@cdo/apps/util/HttpClient';
+
+export async function openaiCompletion(
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/openai.ts]: nit: could we make this function name a bit clearer? something like `get/fetchOpenaiCompletion` or `get/fetchChatCompletion`?

File: apps/src/aichat/openai.ts
Comment (sanchitmalhotra126) [inline apps/src/aichat/openai.ts:1]: nit: can we name this file something like `ChatApi.ts` or similar? On the front-end, we technically don't know we're hitting open AI at all right?

File: apps/src/aichat/openai.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  return await response.json();
+}
+
+const formatForChatCompletion = (chatMessages: ChatCompletionMessage[]) => {
+  return chatMessages.map(message => {
+    return {role: message.role, content: message.chatMessageText};
+  });
+};
+
+type openaiResponse = {role: string; content: string};
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/openai.ts]: nit: maybe `ChatCompletionResponse`?

File: apps/src/aichat/views/ChatWorkspace.tsx [hunk 4]
```diff
@@ -1,35 +1,96 @@
...
+
+    const systemPromptMessage: ChatCompletionMessage = {
+      id: 0,
+      name: 'System',
+      role: Role.SYSTEM,
+      chatMessageText: systemPrompt,
+      status: Status.OK,
+    };
+
+    const messagesToSend = [systemPromptMessage];
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatWorkspace.tsx]: nit: I think this could just be `[systemPromptMessage, ...storedMessages, newMessage]`?

File: apps/src/aichat/views/ChatWorkspace.tsx [hunk 5]
```diff
@@ -1,35 +1,96 @@
...
+import {ChatCompletionMessage} from '../types';
+import {Role, Status} from '../constants';
 import {demoChatMessages} from './demoMessages'; // demo chat messages - remove when connected to backend
+import {openaiCompletion} from '../openai';
 
 const ChatWorkspace: React.FunctionComponent = () => {
+  const [storedMessages, setStoredMessages] =
+    useState<ChatCompletionMessage[]>(demoChatMessages);
+
+  const onSubmit = async (message: string) => {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatWorkspace.tsx:20]: It would be good to break this function up a bit and move the non-React parts into a separate unit (or maybe even into the openai.ts file). One option could be to have a function like `async getChatCompletionMessage(systemPrompt: string, newMessage: string, chatMessages: ChatCompletionMessage[]): Promise<ChatCompletionResponse>` which can handle creating the list of messages and calling the api internally?

File: apps/src/aichat/views/chatWorkspace.module.scss [hunk 6]
```diff
@@ -2,10 +2,11 @@ $message-editor-height: 100px;
 $default-spacing: 2px;
 
 .chatWorkspace {
+    width: 100%;
+    height: 95%;
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/chatWorkspace.module.scss:6]: just curious, why this change?

File: apps/src/aichat/views/ChatWorkspace.tsx (hunk 5 above)
Comment (sanchitmalhotra126) [inline apps/src/aichat/views/ChatWorkspace.tsx:20]: Also, just thinking ahead - this logic actually may be a good candidate for using a [redux async thunk](https://redux-toolkit.js.org/api/createAsyncThunk) once we have our redux store set up. We're changing some state after the submit button is clicked (creating a new 'unknown' chat message) and would probably need to convey some loading information while load is in progress. Doing this in an async thunk would allow us to keep that logic close to redux and simplify the component code. But this is definitely something we can look into later.

**Review by sanchitmalhotra126 — COMMENTED** [review]
This is coming together well! All the logic looks right to me - my comments are mostly around code formatting and organization. Also, it would be great to add docs for the new components (to make sure we're staying on top of documentation early!)

File: apps/src/aichat/views/UserChatMessageEditor.tsx [hunk 7]
```diff
@@ -1,20 +1,29 @@
...
       <Button
         key="submit"
         text="Submit"
         icon="arrow-up"
         onClick={() => {
-          console.log('Submit button clicked');
+          onSubmit?.(userMessage);
+          setUserMessage('');
         }}
         color="purple"
```
Comment (levadadenys) [inline apps/src/aichat/views/UserChatMessageEditor.tsx]: I see it's not a part of this PR, but just something we should not forget about.
We need to use `Button.ButtonColor.brandSecondaryDefault` instead of `here`

File: apps/src/aichat/views/UserChatMessageEditor.tsx [hunk 8]
```diff
@@ -1,20 +1,29 @@
...
-        placeholder="Add a chat message..."
+        placeholder={aichatI18n.userChatMessagePlaceholder()}
+        onChange={e => setUserMessage(e.target.value)}
+        value={userMessage}
       />
       <Button
         key="submit"
         text="Submit"
         icon="arrow-up"
         onClick={() => {
```
Comment (levadadenys) [inline apps/src/aichat/views/UserChatMessageEditor.tsx]: nit: what do you think about wrapping this two buttons' onClicks to `useCallback` hook?

**Review by levadadenys — COMMENTED** [review]
The changes look great to me overall. Only left couple non-blocking comments.
Also just out of curiosity, are we planning to write unit tests for Edubot?

File: apps/src/aichat/openai.ts (hunk 2 above)
Comment (fisher-alice) [reply apps/src/aichat/openai.ts]: Sure! What do you think of `postOpenaiChatCompletion`?

File: apps/src/aichat/openai.ts
Comment (fisher-alice) [reply apps/src/aichat/openai.ts:1]: Sound good - thanks!

File: apps/src/aichat/views/ChatWorkspace.tsx (hunk 4 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx]: Thanks! We'll have to refactor this when we filter out messages from `storedMesages` that are labeled as inappropriate or too personal.

File: apps/src/aichat/openai.ts (hunk 2 above)
Comment (sanchitmalhotra126) [reply apps/src/aichat/openai.ts]: That works!

File: apps/src/aichat/views/ChatWorkspace.tsx (hunk 5 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx:20]: Great idea! Implemented and just pushed. Things are kind of in an interim state since we have pending filters that will be implemented. Lmk what you think - thanks!

File: apps/src/aichat/views/chatWorkspace.module.scss (hunk 6 above)
Comment (fisher-alice) [reply apps/src/aichat/views/chatWorkspace.module.scss:6]: I changed `width` to `100%` because I noticed that if the message is relatively short, the width of the workspace narrows.
I changed `height` to `95%` so the text input box can be seen ene if there is an 'Extra Links' box at the the bottom lefthand corner.

File: apps/src/aichat/views/ChatWorkspace.tsx (hunk 5 above)
Comment (fisher-alice) [reply apps/src/aichat/views/ChatWorkspace.tsx:20]: Okay - that makes sense. @bencodeorg has a branch which is adding a Redux store fo aichat [in this branch.](https://github.com/code-dot-org/code-dot-org/compare/staging...ben/openai-fe-client) I'll connect with him when he's back about this idea - thanks!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 7 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: Thanks! Here's a screenshot: 
Screen Shot 2023-08-14 at 4 48 03 PM

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 8 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: I don't have a preference. @sanchitmalhotra126 - thoughts?

**Comment by fisher-alice** [comment]
> The changes look great to me overall. Only left couple non-blocking comments. Also just out of curiosity, are we planning to write unit tests for Edubot?

Thanks Denys! When the AI Chat eng team meets again, I definitely want to discuss plans/guidelines for testing and documentation.

**Comment by fisher-alice** [comment]
> This is coming together well! All the logic looks right to me - my comments are mostly around code formatting and organization. Also, it would be great to add docs for the new components (to make sure we're staying on top of documentation early!)

Thanks @sanchitmalhotra126! What kind of documentation are you thinking of? I can get started on that but would love to see an example. Thanks!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 7 above)
Comment (levadadenys) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: That's awesome, thank you!

**Comment by levadadenys** [comment]
> > This is coming together well! All the logic looks right to me - my comments are mostly around code formatting and organization. Also, it would be great to add docs for the new components (to make sure we're staying on top of documentation early!)
> 
> Thanks @sanchitmalhotra126! What kind of documentation are you thinking of? I can get started on that but would love to see an example. Thanks!

Not sure if @sanchitmalhotra126 meant this, but want to add from myself and design system work experience - we can create storybook documentation for specific component(s) (for examples see `apps/src/componentLibrary/checkbox/Checkbox.tsx` or any other design system component in `/componentLibrary`)
If needed - it's also possible to create Readme.md / Changelog.md for `apps/src/aichat/views` or `apps/src/aichat`

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 8 above)
Comment (sanchitmalhotra126) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: Doesn't hurt! We should probably be using `useCallback` more in general, thanks Denys for keeping us on it :)

**Comment by sanchitmalhotra126** [comment]
> Thanks @sanchitmalhotra126! What kind of documentation are you thinking of? I can get started on that but would love to see an example. Thanks!

Oh I was just referring to top level comments in new components/functions. I think it would be good to at least have a description on every top level view component file, and for the API files well. See Lab2 and Music Lab, I think we have docs on most of those classes.

File: apps/src/aichat/chatApi.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+export async function postOpenaiChatCompletion(
+  chatMessages: ChatCompletionMessage[]
+): Promise<ChatCompletionResponse> {
+  const payload = {messages: formatForChatCompletion(chatMessages)};
+  // Send request to chat completion backend controller.
+  const url = '/openai/chat_completion';
+  const response = await HttpClient.post(url, JSON.stringify(payload), true, {
+    'Content-Type': 'application/json; charset=UTF-8',
+  });
+  if (response.status === 200) {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: We can also just check `response.ok` here

File: apps/src/aichat/chatApi.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
+import {ChatCompletionMessage} from './types';
+import HttpClient from '@cdo/apps/util/HttpClient';
+import {Role, Status} from './constants';
+
+export async function postOpenaiChatCompletion(
+  chatMessages: ChatCompletionMessage[]
+): Promise<ChatCompletionResponse> {
+  const payload = {messages: formatForChatCompletion(chatMessages)};
+  // Send request to chat completion backend controller.
+  const url = '/openai/chat_completion';
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: minor: my preference is to keep hardcoded paths like these at the top of the file for visibility, but up to you

File: apps/src/aichat/chatApi.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import {ChatCompletionMessage} from './types';
+import HttpClient from '@cdo/apps/util/HttpClient';
+import {Role, Status} from './constants';
+
+export async function postOpenaiChatCompletion(
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: Does this need to be exported? Looks like it's only used internally right?

File: apps/src/aichat/chatApi.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+};
+
+export async function getChatCompletionMessage(
+  systemPrompt: string,
+  lastMessageId: number,
+  newMessage: string,
+  chatMessages: ChatCompletionMessage[]
+): Promise<ChatCompletionMessage[]> {
+  const newChatMessage: ChatCompletionMessage = {
+    id: lastMessageId + 1,
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: Does `postOpenaiChatCompletion` need to know about other fields like id, name, status, etc, other than `role` and `chatMessageText`? I wonder if it would be better to instead convert the newMessage and systemPrompt into the format we need for the API directly, rather than first converting them into a `ChatCompletionMessage` and then into the API format. That way we're not doing any extraneous ID management and wrapping inside this function.

File: apps/src/aichat/chatApi.ts [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+    id: newChatMessage.id + 1,
+    name: Role.ASSISTANT, // Will be reassigned in ChatWorkspace.tsx.
+    role: Role.ASSISTANT,
+    status: Status.OK,
+    chatMessageText: assistantMessage,
+  };
+  return [newChatMessage, assistantChatMessage];
+}
+
+type ChatCompletionResponse = {role: string; content: string} | null;
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: nit: probably better to remove the `null` here and make the return type of the function `ChatCompletionResponse | null` so it's clearer

File: apps/src/aichat/constants.ts [hunk 14]
```diff
@@ -0,0 +1,12 @@
+export enum Role {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/constants.ts]: We can probably just move these into `types`

File: apps/src/aichat/chatApi.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    return {role: message.role, content: message.chatMessageText};
+  });
+};
+
+export async function getChatCompletionMessage(
+  systemPrompt: string,
+  lastMessageId: number,
+  newMessage: string,
+  chatMessages: ChatCompletionMessage[]
+): Promise<ChatCompletionMessage[]> {
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: Might be good to be more specific about the response type here. What about something like
```
{
   status: 'ok' | 'inappropriate' | ...etc
   assistantResponse?: string
}
```
With the understanding that if status is not `ok`, then `assistantResponse` will be undefined. This way, this API component doesn't need to know too much about how to form a `ChatCompletionMessage` and managing additional fields like id, name, etc.

File: apps/src/aichat/chatApi.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    newChatMessage.status = Status.INAPPROPRIATE; // TODO: Update status to be more accurate.
+    return [newChatMessage];
+  }
+  // Response was successful, so update status of new message.
+  newChatMessage.status = Status.OK;
+
+  const assistantMessage = response.content;
+  const assistantChatMessage: ChatCompletionMessage = {
+    id: newChatMessage.id + 1,
+    name: Role.ASSISTANT, // Will be reassigned in ChatWorkspace.tsx.
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: Does `name` have to be part of this interface at all actually? Looks like we get it from the level directly, so maybe it's easier to just pass it directly to the view instead of needing to set it here and reassign it?

**Review by sanchitmalhotra126 — COMMENTED** [review]
Feedback mostly on the new API class - coming along well!

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 8 above)
Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: Thanks to you both!

Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: Do one of you mind explaining why we should be using `useCallback` more?

File: apps/src/aichat/chatApi.ts (hunk 10 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: Added a constant to `constants.ts`.

File: apps/src/aichat/chatApi.ts (hunk 12 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: Yeah - that makes sense. Will update.

File: apps/src/aichat/chatApi.ts (hunk 15 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: Sounds good - thank you!

File: apps/src/aichat/chatApi.ts (hunk 9 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: Cool - thanks!

File: apps/src/aichat/chatApi.ts (hunk 16 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: I agree - it is definitely redundant for user and system.

File: apps/src/aichat/chatApi.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  newMessage: string,
+  chatMessages: ChatCompletionMessage[]
+): Promise<ChatCompletionResponse> {
+  const messagesToSend = [
+    {role: Role.SYSTEM, content: systemPrompt},
+    ...formatForChatCompletion(chatMessages),
+    {role: Role.USER, content: newMessage},
+  ];
+
+  const response = await postOpenaiChatCompletion(messagesToSend);
```
Comment (sanchitmalhotra126) [inline apps/src/aichat/chatApi.ts]: Oh one thing I forgot - we should wrap this in a `try`/`catch` since the HttpClient will throw on any errors. For now we could just console log, or we could go ahead and report an error using `Lab2MetricsReporter`, but it would be good to at least have some error handling here.

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks good! Thanks for all the iteration on this!

**Comment by fisher-alice** [comment]
> Looks good! Thanks for all the iteration on this!

Thank you for your feedback and guidance!

File: apps/src/aichat/chatApi.ts (hunk 17 above)
Comment (fisher-alice) [reply apps/src/aichat/chatApi.ts]: Thanks! Added a `try/catch`. In terms of conveying error to view - will add a ticket for this.
Just adding notes: once PII and content filters are in place, if the user message is too personal (`Status.PERSONAL`), the message will not be displayed and instead a warning with a yellow background will show up.
If inappropriate (`Status.INAPPROPRIATE`), a warning with a red background will be displayed.
But if the post request just fails due to a timeout or if chat assistant response is inappropriate (I think we'll eventually monitor this) - need to decide how to handle these cases.

File: apps/src/aichat/views/UserChatMessageEditor.tsx (hunk 8 above)
Comment (levadadenys) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: @fisher-alice sorry for long response.
Sure, `useCallback` memoizes callback functions, so instead of creating a new callback function on every component's render - we'll have one callback function across multiple renders(untill one of callback's dependencies'll change), meaning perfomance optimization.
Here're couple pages about `useCallback` if you're interested: https://www.w3schools.com/react/react_usecallback.asp
https://www.freecodecamp.org/news/better-react-performance-usecallback-vs-usememo/

Also, in my opinion, using useCallback makes code cleaner by separating views(jsx,props,etc) and logic parts of code(callbacks, parsing data/props, etc)

Comment (fisher-alice) [reply apps/src/aichat/views/UserChatMessageEditor.tsx]: Thank you for the explanation!
